### ISSUE TITLE
Can one of this autofocus methods can be used for non-satellite SAR data

### ISSUE THREAD

**Issue by PalgunaGopireddy** [aisari/AutofocusSAR#7]
Hi.
I have a SAR dataset. So I have `SI` value of the dataset, No  `ca`, `cr` values available. 
Can I use one the autofocus methods be used for this dataset. If so which one?

**Comment by aisari** [comment]
If the phase errors can be modeled as polynomial, both the ECELMs and PAFnet methods can be used. Otherwise, if the type of phase errors is not known, you can use VDNNAF.  However, both of the above methods need to be retrained to fit your SAR platform to get the best focusing quality.

Note: You should have well focused ground-truth to train the VDNNAF model. Our proposed PAFnet method is unsupervised or self-supervised, replace the polynomial to your phase error type model to correct new type of errors.

**Comment by PalgunaGopireddy** [comment]
Thanks. Let me run that and will update if any issue.

**Comment by PalgunaGopireddy** [comment]
Hi, I ran `VDNNAF` model `train.py` and `test.py`

In the test.py used './record/PE1d/weights/ValidSingleAverage204.pth.tar'
But the `train.py` did not create this file. I downloaded it from the `https://drive.google.com/drive/folders/1c3eXxOVtGvX90w2zTyY1ITVyTSlN9l69?usp=sharing`

As I will be training the model with my dataset 
Can I use use the `/snapshot/VDNN/SimPoly/AdamW/2021/weights/TestSingle1.pth` in that place?

Are these both `.pth` files same. Else how can I create `ValidSingleAverage204.pth.tar`

**Comment by aisari** [comment]
To save the disk space, the weights that meet with the following conditions will be saved:

1. the weights that has the lowest loss --> DatasetSingleXXX.pth.tar
2. the weights that has the lowest average loss of the last 50 epoch --> DatasetAverageXXX.pth.tar
3. both 1 and 2 --> DatasetSingleAverageXXX.pth.tar

where, XXX means the number of epochs that have been trained.



---- Replied Message ----
| From | ***@***.***> |
| Date | 11/29/2023 21:07 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [aisari/AutofocusSAR] Can one of this autofocus methods can be used for non-satellite SAR data (Issue #7) |

Hi, I ran VDNNAF model train.py and test.py

In the test.py used './record/PE1d/weights/ValidSingleAverage204.pth.tar'
But the train.py did not create this file. I downloaded it from the https://drive.google.com/drive/folders/1c3eXxOVtGvX90w2zTyY1ITVyTSlN9l69?usp=sharing

As I will be training the model with my dataset
Can I use use the /snapshot/VDNN/SimPoly/AdamW/2021/weights/TestSingle1.pth in that place?

Are these both .pth files same. Else how can I create ValidSingleAverage204.pth.tar

**Comment by aisari** [comment]
The ``losslog.judge`` function in the train.py performs the above rules. The above rules are not mandatory, you can  save the weights of each epoch, just select the weights that has the best performance on the validation dataset.



---- Replied Message ----
| From | ***@***.***> |
| Date | 11/29/2023 21:07 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [aisari/AutofocusSAR] Can one of this autofocus methods can be used for non-satellite SAR data (Issue #7) |

Hi, I ran VDNNAF model train.py and test.py

In the test.py used './record/PE1d/weights/ValidSingleAverage204.pth.tar'
But the train.py did not create this file. I downloaded it from the https://drive.google.com/drive/folders/1c3eXxOVtGvX90w2zTyY1ITVyTSlN9l69?usp=sharing

As I will be training the model with my dataset
Can I use use the /snapshot/VDNN/SimPoly/AdamW/2021/weights/TestSingle1.pth in that place?

Are these both .pth files same. Else how can I create ValidSingleAverage204.pth.tar

**Comment by aisari** [comment]
You can not use TestSingle1.path.tar since the weights in this file only be updated one epoch. If you use, the model will perform very bad.



---- Replied Message ----
| From | ***@***.***> |
| Date | 11/29/2023 21:07 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [aisari/AutofocusSAR] Can one of this autofocus methods can be used for non-satellite SAR data (Issue #7) |

Hi, I ran VDNNAF model train.py and test.py

In the test.py used './record/PE1d/weights/ValidSingleAverage204.pth.tar'
But the train.py did not create this file. I downloaded it from the https://drive.google.com/drive/folders/1c3eXxOVtGvX90w2zTyY1ITVyTSlN9l69?usp=sharing

As I will be training the model with my dataset
Can I use use the /snapshot/VDNN/SimPoly/AdamW/2021/weights/TestSingle1.pth in that place?

Are these both .pth files same. Else how can I create ValidSingleAverage204.pth.tar

**Comment by PalgunaGopireddy** [comment]
Hi. Thanks for information.

n1 we choose the number of epochs, we want the lowest average for.
I am running with your dataset to understand it, so I am running it for like 8 epochs, so I choose `n1` to be 4 epochs for my case. 
But, I can't understand the meaning of `n2` in `losslog.judge`. What is it?

**Comment by PalgunaGopireddy** [comment]
You said _`To save the disk space, the weights that meet with the following conditions will be saved:`_ and _`The ``losslog.judge`` function in the train.py performs the above rules. The above rules are not mandatory`_

So I guess you did not provide code to save `ValidSingleAverage.pth.tar` file. as after running `train.py` file did not save any `ValidSingleAverage.pth.tar` file.

So Can I use `ValidSingleXXX.pth.tar`? in `test.py`

You also told _`not to use TestSingle1.path.tar`._

**Comment by antsfamily** [comment]
If no ValidSingleAverage.pth.tar file has been saved, please check the validation loss curve. The losses should gradually decrease. If the average loss of the last n2 epochs is less than the average of the previous n1 epochs and the loss value of the current epoch is the smallest among the n1 epochs, then save the weights of current epoch with Average flag. The following two graphs show the loss curves of AFnet and VDNNAF.

![Loss_Epoch](https://github.com/aisari/AutofocusSAR/assets/11564731/85e61426-01b2-4bf8-9242-6a1450c9a517)

![VDNNAF](https://github.com/aisari/AutofocusSAR/assets/11564731/191aaa45-3bc8-4280-9a07-3b3de3dfd03d)

**Comment by PalgunaGopireddy** [comment]
Hi. The images you uploaded are not visible. 
Here is the loss curve saved in `./snapshot/VDNN/SimPoly/AdamW/2021` . I have taken n1 =8, n2 =4 for this and ran for 8 epochs
![LossEpoch](https://github.com/aisari/AutofocusSAR/assets/106730853/8b5bb2f0-2014-4071-97ef-b10ac60eae17)

**Comment by PalgunaGopireddy** [comment]
Also, could you tell where the input-defocused and focused images are saved? As I see  `./snapshot/VDNN/SimPoly/AdamW/2021/images` has both the focused, defocued and ground truth images. Are the focused images saved after the run of `test.py`

**Comment by antsfamily** [comment]
> Hi. The images you uploaded are not visible. Here is the loss curve saved in `./snapshot/VDNN/SimPoly/AdamW/2021` . I have taken n1 =8, n2 =4 for this and ran for 8 epochs ![LossEpoch](https://private-user-images.githubusercontent.com/106730853/288966821-8b5bb2f0-2014-4071-97ef-b10ac60eae17.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIwMTgwMjMsIm5iZiI6MTcwMjAxNzcyMywicGF0aCI6Ii8xMDY3MzA4NTMvMjg4OTY2ODIxLThiNWJiMmYwLTIwMTQtNDA3MS05N2VmLWIxMGFjNjBlYWUxNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIwOFQwNjQyMDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xOWY5NzYyODI5NzBmZTc1NjA4ZTlkZmFlNzFkZDczMWJhN2MzODZhNGUzNmYxODI5MjgwMmQ0OTcwOGQwNjEwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.LpD003NyNbZJisEfJeLY9YHGFYwAA5SbXe6cNaddF_U)

You should run at least n1 + n2 epochs. You can run more than 100 epochs.

**Comment by antsfamily** [comment]
> Also, could you tell where the input-defocused and focused images are saved? As I see `./snapshot/VDNN/SimPoly/AdamW/2021/images` has both the focused, defocued and ground truth images. Are the focused images saved after the run of `test.py`

You should check ``issaveimage`` flag.

**Comment by PalgunaGopireddy** [comment]
> > Hi. The images you uploaded are not visible. Here is the loss curve saved in `./snapshot/VDNN/SimPoly/AdamW/2021` . I have taken n1 =8, n2 =4 for this and ran for 8 epochs ![LossEpoch](https://private-user-images.githubusercontent.com/106730853/288966821-8b5bb2f0-2014-4071-97ef-b10ac60eae17.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIwMTgwMjMsIm5iZiI6MTcwMjAxNzcyMywicGF0aCI6Ii8xMDY3MzA4NTMvMjg4OTY2ODIxLThiNWJiMmYwLTIwMTQtNDA3MS05N2VmLWIxMGFjNjBlYWUxNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIwOFQwNjQyMDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xOWY5NzYyODI5NzBmZTc1NjA4ZTlkZmFlNzFkZDczMWJhN2MzODZhNGUzNmYxODI5MjgwMmQ0OTcwOGQwNjEwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.LpD003NyNbZJisEfJeLY9YHGFYwAA5SbXe6cNaddF_U)
> 
> You should run at least n1 + n2 epochs. You can run more than 100 epochs.

Thanks. I tried running for the default 800 epochs, Due to having only 32 GB RAM device it took 1 day for 20 epochs. that's why I had to change n1 and n2 and epoch values.

**Comment by PalgunaGopireddy** [comment]
> > Also, could you tell where the input-defocused and focused images are saved? As I see `./snapshot/VDNN/SimPoly/AdamW/2021/images` has both the focused, defocued and ground truth images. Are the focused images saved after the run of `test.py`
> 
> You should check `issaveimage` flag.

I will try and see and update here.

**Comment by PalgunaGopireddy** [comment]
> > Also, could you tell where the input-defocused and focused images are saved? As I see `./snapshot/VDNN/SimPoly/AdamW/2021/images` has both the focused, defocued and ground truth images. Are the focused images saved after the run of `test.py`
> 
> You should check `issaveimage` flag.

 I came to know that running `train.py` saves random images of the defocused, ground-truth, focused `train`, `valid`, and `test` dataset images for visual verification. In every epoch, the focused image changes better. So the more the number of epochs more the number of focusing.

**Comment by aisari** [comment]
You should also check the "outfolder" variable in test.py, it specifies the output folder.

Generally speaking, the more the better, however, you should check the loss on validation dataset to avoid overfitting.



| |
***@***.***
|
|
***@***.***
|




---- Replied Message ----
| From | ***@***.***> |
| Date | 12/09/2023 21:10 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [aisari/AutofocusSAR] Can one of this autofocus methods can be used for non-satellite SAR data (Issue #7) |

Also, could you tell where the input-defocused and focused images are saved? As I see ./snapshot/VDNN/SimPoly/AdamW/2021/images has both the focused, defocued and ground truth images. Are the focused images saved after the run of test.py

You should check issaveimage flag.

I came to know that running train.py saves random images of the defocused, ground-truth, focused train, valid, and test dataset images for visual verification. In every epoch, the focused image changes better. So the more the number of epochs more the number of focusing.
